src/track/macro.cpp

```diff
@@ -0,0 +1,151 @@
+#include "track/macro.h"
+
+#include <QDebug>
+
+#include "util/assert.h"
+
+proto::Macro_Action* MacroAction::serialize() const {
+    auto serialized = new proto::Macro_Action();
+    serialized->set_position(position);
+    serialized->set_target(target);
+    serialized->set_type(static_cast<uint32_t>(type));
+    return serialized;
+}
+
+Macro::Macro(QList<MacroAction> actions, QString label, State state, int dbId)
+        : m_bDirty(false),
+          m_iId(dbId),
+          m_actions(actions),
+          m_label(label),
+          m_state(state) {
+}
+
+// static
+QList<MacroAction> Macro::deserialize(QByteArray serialized) {
+    proto::Macro macroProto = proto::Macro();
+    macroProto.ParseFromArray(serialized.data(), serialized.length());
+    QList<MacroAction> result;
+    result.reserve(macroProto.actions_size());
+    for (const proto::Macro_Action& action : macroProto.actions()) {
+        result.append(MacroAction(action.position(), action.target()));
+    }
+    return result;
+}
+
+// static
+QByteArray Macro::serialize(QList<MacroAction> actions) {
```

Comment:
```suggestion
QByteArray Macro::serialize(const QList<MacroAction>& actions) {
```

Reply:
implicit sharing?

Reply:
This requires some extra cpu cycles if the object is not yet shared. 
Because we don't rely on that we pass Qt containers always as const reference, except it is stored internally.

Reply:
What about the return types?

Reply:
That the same, if you return a member variable use a const reference. If you return a local, return a copy. The copy is likely omitted in this case due to return value optimization.